Hi Jason, I played a bit with my code and I have to push back a bit on delegating the type checking logic to the C++ compiler.

I tried to introduce a casting error in my code and the error log was not obscure at all!
```

CHI-dvm-misc-node-actions.sm:217: Warning: Non-void return ignored, return type is 'Tick'
CHI-dvm-misc-node-actions.sm:221: Warning: Non-void return ignored, return type is 'Tick'
 [VER TAGS]  -> ARM/sim/tags.cc
 [     CXX] ARM/mem/ruby/protocol/MiscNode_Controller.py.cc -> .o
 [     CXX] ARM/mem/ruby/protocol/MiscNode_Controller.py.cc -> .pyo
 [     CXX] ARM/mem/ruby/protocol/Memory_Controller.py.cc -> .pyo
 [     CXX] ARM/mem/ruby/protocol/Memory_Controller.py.cc -> .o
 [     CXX] ARM/mem/ruby/protocol/Cache_Controller.cc -> .o
build/ARM/mem/ruby/protocol/Cache_Controller.cc: In member function ‘gem5::ruby::Cache_TBE* gem5::ruby::Cache_Controller::allocateRequestTBE(const Addr&, const gem5::ruby::CHIRequestMsg&)’:
build/ARM/mem/ruby/protocol/Cache_Controller.cc:9690:26: error: invalid ‘static_cast’ from type ‘gem5::ruby::Cache_TBE*’ to type ‘gem5::Addr’ {aka ‘long unsigned int’}
 9690 |         (*tbe).m_txnId = static_cast<Addr >(tbe)
      |                          ^~~~~~~~~~~~~~~~~~~~~~~
scons: *** [build/ARM/mem/ruby/protocol/Cache_Controller.o] Error 1
```

The only disadvantage of it is that it does not reference the protocol file/line directly so a developer would have to infer it from the name of the function/method. It doesn't really seem to me a big deal. Considering that you are happy to allow casting between integer types, the chance of someone making mistakes when casting the few custom types usable in SLICC is really minimal... Please let me know your thoughts :)

--- a/src/mem/ruby/protocol/chi/CHI-cache-actions.sm
+++ b/src/mem/ruby/protocol/chi/CHI-cache-actions.sm
@@ -2567,6 +2567,8 @@ action(Send_CompI, desc="") {
     out_msg.type := CHIResponseType:Comp_I;
     out_msg.responder := machineID;
     out_msg.Destination.add(tbe.requestor);
+    out_msg.Destination.add(tbe.requestor);
+    out_msg.txnId := tbe.txnId;
     out_msg.dbid := tbe.txnId;
   }
 }

--- a/src/mem/ruby/protocol/chi/CHI-cache-actions.sm
+++ b/src/mem/ruby/protocol/chi/CHI-cache-actions.sm
@@ -2644,6 +2644,7 @@ action(Send_CompDBIDResp_Stale, desc="") {
     out_msg.type := CHIResponseType:CompDBIDResp;
     out_msg.responder := machineID;
     out_msg.Destination.add(tbe.requestor);
+    out_msg.txnId := tbe.txnId;
     out_msg.dbid := tbe.txnId;
     // We don't know if this is a stale writeback or a bug, so flag the
     // reponse so the requestor can make further checks

